File: samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/actors/PlayActorRelocationAdapter.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.actors;

import com.fasterxml.jackson.databind.ObjectMapper;

import systems.zlink.framework.actors.ZLinkActorRelocationAdapter;
import systems.zlink.framework.actors.ZLinkRelocationCancellation;
import systems.zlink.samples.tictactoe.shared.contracts.PlayerInfo;

import java.io.IOException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

// --8<-- [start:doc-relocation-adapter]
public final class PlayActorRelocationAdapter implements ZLinkActorRelocationAdapter<PlayActor> {
    private static final ObjectMapper JSON = new ObjectMapper();

    // --8<-- [start:doc-ttt-actor-capture]
    @Override
    public CompletionStage<byte[]> capture(
            PlayActor actor, ZLinkRelocationCancellation cancellation) {
        try {
            return CompletableFuture.completedFuture(
                    JSON.writeValueAsBytes(
                            new TransferState(
                                    actor.joinedRoomId(),
                                    actor.playerOrNull(),
                                    actor.destroyAfterEntrySpotJoin(),
                                    actor.disconnected())));
        } catch (IOException error) {
            return CompletableFuture.failedFuture(error);
        }
    }

    // --8<-- [end:doc-ttt-actor-capture]

    @Override
    public CompletionStage<Void> restore(
            PlayActor actor, byte[] state, ZLinkRelocationCancellation cancellation) {
        try {
            TransferState transferred = JSON.readValue(state, TransferState.class);
            if (transferred.player() != null) {
                actor.applyPlayer(transferred.player());
            }
            if (transferred.roomId() != null && !transferred.roomId().isBlank()) {
                actor.joinGame(transferred.roomId());
            }
            if (transferred.destroyAfterEntrySpotJoin()) {
                actor.markForDestroyAfterRoomLeave();
            }
            if (transferred.disconnected()) {
                actor.markDisconnected();
            }
            return CompletableFuture.completedFuture(null);
        } catch (IOException error) {
            return CompletableFuture.failedFuture(error);
        }
    }

    public record TransferState(
            String roomId,
            PlayerInfo player,
            boolean destroyAfterEntrySpotJoin,
            boolean disconnected) {}
}
// --8<-- [end:doc-relocation-adapter]

File: samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/infrastructure/zlink/actors/PlayActorRelocationAdapter.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.actors

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import systems.zlink.framework.actors.ZLinkActorRelocationAdapter
import systems.zlink.framework.actors.ZLinkRelocationCancellation
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerInfo

// --8<-- [start:doc-relocation-adapter]
class PlayActorRelocationAdapter : ZLinkActorRelocationAdapter<PlayActor> {
    // --8<-- [start:doc-ttt-actor-capture]
    override fun capture(
        actor: PlayActor,
        cancellation: ZLinkRelocationCancellation,
    ): CompletionStage<ByteArray> =
        CompletableFuture.completedFuture(
            json.writeValueAsBytes(
                TransferState(
                    actor.joinedRoomIdOrNull(),
                    actor.playerOrNull(),
                    actor.destroyAfterEntrySpotJoin,
                    actor.disconnected,
                )
            )
        )

    // --8<-- [end:doc-ttt-actor-capture]

    override fun restore(
        actor: PlayActor,
        state: ByteArray,
        cancellation: ZLinkRelocationCancellation,
    ): CompletionStage<Void> {
        val transferred = json.readValue(state, TransferState::class.java)
        transferred.player?.let(actor::applyPlayer)
        transferred.roomId?.takeIf(String::isNotBlank)?.let(actor::joinGame)
        if (transferred.destroyAfterEntrySpotJoin) actor.markForDestroyAfterRoomLeave()
        if (transferred.disconnected) actor.markDisconnected()
        return CompletableFuture.completedFuture(null)
    }

    companion object {
        private val json = jacksonObjectMapper()
    }

    data class TransferState(
        val roomId: String?,
        val player: PlayerInfo?,
        val destroyAfterEntrySpotJoin: Boolean,
        val disconnected: Boolean,
    )
}
// --8<-- [end:doc-relocation-adapter]
